Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/performance/AzurePerformanceMetricsCollector.java

```diff
@@ -0,0 +1,142 @@
+ *
+ * All rights reserved.
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+ * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+ * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+ * the Software.
+ *
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+ * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.microsoft.azure.toolkit.lib.common.performance;
+
+import com.microsoft.azure.toolkit.lib.common.operation.AzureOperation;
+import com.microsoft.azure.toolkit.lib.common.operation.AzureOperationRef;
+import com.microsoft.azure.toolkit.lib.common.operation.AzureOperationUtils;
+import com.microsoft.azure.toolkit.lib.common.task.AzureTask;
+import com.microsoft.azure.toolkit.lib.common.task.AzureTaskContext;
+import com.microsoft.azuretools.azurecommons.helpers.NotNull;
+import com.microsoft.azuretools.telemetry.AppInsightsClient;
+import lombok.SneakyThrows;
+
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class AzurePerformanceMetricsCollector {
+
+    public static final String NONE = "NONE";
+    public static final String PERFORMANCE = "PERFORMANCE";
+
+    public static void beforeEnter(final AzureOperationRef operation) {
+        final Deque<AzureOperationRef> operations = AzureTaskContext.getContextOperations();
+        AzurePerformanceMetricsCollector.enterOperation(operation, operations);
+    }
+
+    public static void afterExit(final AzureOperationRef operation) {
+        final Deque<AzureOperationRef> operations = AzureTaskContext.getContextOperations();
+        AzurePerformanceMetricsCollector.exitOperation(operation, operations);
+    }
+
+    public static void beforeEnter(final AzureTask<?> task) {
+        AzurePerformanceMetricsCollector.enterTask(task);
+    }
+
+    public static void afterExit(final AzureTask<?> task) {
+        AzurePerformanceMetricsCollector.exitTask(task);
+        final AzureTaskContext.Node context = task.getContext();
+        final Deque<AzureOperationRef> operations = AzureTaskContext.getContextOperations(context);
+        while (!operations.isEmpty()) { // end operations asynchronously
```

Comment:
why we need to end a lot of operations? end current operation is enough?

Reply:
this is to actually make it fast/fast in later summarizing, it's no longer needed with the new id schema.